cluster/cluster_members.go

```diff
@@ -270,3 +270,56 @@ func GetClusterMemberAPIExtensions(ctx context.Context, tx *sql.Tx) ([]extension
 
 	return results, nil
 }
+
+// GetUpgradingClusterMembers returns the list of all cluster members during an upgrade, as well as a map of members who we consider to be in a waiting state.
+// This function can be used immediately after dqlite is ready, before we have loaded any prepared statements.
+func GetUpgradingClusterMembers(ctx context.Context, tx *sql.Tx, schemaInternal uint64, schemaExternal uint64, apiExtensions extensions.Extensions) (allMembers []InternalClusterMember, awaitingMembers map[string]bool, err error) {
+	tableName, err := prepareUpdateV1(tx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check for the `api_extensions` column, which may not exist if we haven't actually run the update yet.
+	stmt := fmt.Sprintf(`
+SELECT count(name)
+FROM pragma_table_info('%s')
+WHERE name IN ('api_extensions');
+`, tableName)
+
+	var count int
+	err = tx.QueryRow(stmt).Scan(&count)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Fetch all cluster members with a smaller schema version than we expect.
+	stmt = `SELECT id, name, address, certificate, schema_internal, schema_external, %s, heartbeat, role
+  FROM %s
+  ORDER BY name
+	`
+
+	// If API extensions are supported, ensure the list for each cluster member also matches what we expect,
+	// and only return cluster members for whom it does not.
+	apiField := "'[]' as api_extensions"
+	if count == 1 {
+		apiField = "api_extensions"
+	}
+
+	stmt = fmt.Sprintf(stmt, apiField, tableName)
```

Comment:
To not open any doors for injections in the future, should `tableName` be a prepared statement instead? 

Right now that is not an issue as `prepareUpdateV1` returns only defined strings or `internal_cluster_members_new` if it's found in the DB.

Reply:
Hmm, I didn't know you could do that with the table name as well. I'll give it a try.

Reply:
Sadly, this doesn't work in sqlite. You get `Error: near "?": syntax error`

Reply:
Ah yeah that's true. Don't want to over complicate things now but maybe what we have to do in the future is check against a whitelist of allowed table names?